dill/session.py

```diff
@@ -34,6 +57,54 @@
 
 TEMPDIR = pathlib.PurePath(tempfile.gettempdir())
 
+class _PeekableReader:
+    """lightweight readable stream wrapper that implements peek()"""
+    def __init__(self, stream):
+        self.stream = stream
+    def read(self, n):
+        return self.stream.read(n)
+    def readline(self):
+        return self.stream.readline()
+    def tell(self):
+        return self.stream.tell()
+    def close(self):
+        return self.stream.close()
+    def peek(self, n):
+        stream = self.stream
+        try:
+            if hasattr(stream, 'flush'): stream.flush()
+            position = stream.tell()
+            stream.seek(position)  # assert seek() works before reading
+            chunk = stream.read(n)
+            stream.seek(position)
+            return chunk
+        except (AttributeError, OSError):
+            raise NotImplementedError("stream is not peekable: %r", stream) from None
+
+def _open(file, mode, *, peekable=False):
+    """return a context manager with an opened file-like object"""
+    import io
+    attr = 'write' if 'w' in mode else 'read'
```

Comment:
what about "write" file modes: `x`, `a`, and `r+`?

Reply:
changed to `readonly = ('r' in mode and '+' not in mode)`